File: Data Structures & Algorithms/capacity-to-ship-packages-within-d-days/submission-1.java
class Solution {
    public int shipWithinDays(int[] weights, int days) {
        int[] bounds = stats(weights);
        int left = bounds[0];
        int right = bounds[1];
        while (left <= right) {
            int mid = (left + right) / 2;
            boolean works = possible(weights, days, mid);
            if (works) {
                right = mid - 1;
            } else {
                left = mid + 1;
            }
            System.out.println(mid + "=" + works);
        }
        return left;
    }

    private boolean possible(int[] weights, int days, int weight) {
        int count = 1;
        int running = 0;
        for (int i = 0; i < weights.length; i++) {
            running += weights[i];
            if (running > weight) {
                count++;
                running = 0;
                i--;
                if (count > days) return false;
            }
        }
        return true;
    }

    private int[] stats(int[] weights) {
        int max = 0;
        int sum = 0;
        for (int weight : weights) {
            max = Math.max(max, weight);
            sum += weight;
        }
        return new int[] {max, sum};
    }
}

File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-1.cpp
class Solution {
public:
    int characterReplacement(string s, int k) {
        int left = 0;
        int arr[26] = {0};
        int maxCount = 0;
        int ans = 0;
        for (int right = 0; right < s.size(); right++) {
            arr[s[right] - 'A']++;
            maxCount = max(maxCount, arr[s[right] - 'A']);
            while ((right - left + 1) - maxCount > k) {
                arr[s[left] - 'A']--;
                left++;
            }
            ans = max(ans, right - left + 1);
        }
        return ans;
    }
};
